### PR TITLE
feat: mret to main in supervisor mode

### LINKED ISSUE #6: Switch from M-mode to S-mode after Kernel bootup

### PR DESCRIPTION
**Pull request by bluebrown**
Fixes #6

### CODE REVIEW

File: linker.ld [hunk 1]
```diff
@@ -1,44 +1,40 @@
...
   .data : ALIGN(4K) {
-    *(.data .data.*)
-  }
+    *(.sdata .sdata*);
+    *(.data .data*);
+  } > DRAM
 
   .bss : ALIGN(4K) {
-    *(.bss .bss.*)
-    *(.sbss* .sbss.*)
```
Comment (bluebrown) [inline linker.ld:40]: maybe we should ass sbss back. I havent noticed its missed yet, but may be important at some point.

File: .cargo/config.toml [hunk 2]
```diff
@@ -2,12 +2,9 @@
...
+rustflags = [ "-C", "code-model=medium"]
 runner = """
-  qemu-system-riscv64 \
-    -machine virt \
-    -bios none \
-    -smp 1 \
-    -nographic \
-    -kernel
+  qemu-system-riscv64 -machine virt
+    -display none -serial stdio -smp 2
```
Comment (Archisman-Mridha) [inline .cargo/config.toml:8]: Can we just stick to a single CPU core for now?

Comment (Archisman-Mridha) [reply .cargo/config.toml:8]: And we should add [this](https://docs.amd.com/r/en-US/ug1304-versal-acap-ssdg/Symmetric-Multiprocessing) as a reference for someone who wants to understand - `what's SMP`.

File: build.rs [hunk 3]
```diff
@@ -1,4 +1,15 @@
 fn main() {
   println!("cargo:rerun-if-changed=linker.ld");
   println!("cargo:rerun-if-changed=src/asm/entry.S");
+  println!("cargo:rustc-link-arg=-Tlinker.ld");
+
+  // rust's build system sucks. I want to do this, but
+  // the cc, isnt configured the same. It uses a different ABI.
+  //
+  //     cc::Build::new().file("src/asm/entry.S").compile("entry");
```
Comment (Archisman-Mridha) [inline build.rs:9]: It's fine.

We have `core::arch::global_asm!(include_str!("asm/entry.S"));` right before the `start` function in `main.rs`. It's helpful for visualizing the execution flow :

- We start execution from the `entry` symbol in `entry.S` file.
- Then we enter the `start` function.

File: linker.ld
Comment (Archisman-Mridha) [inline linker.ld:1]: Why did you remove all the comments :). Can we please keep them.

Comment (Archisman-Mridha) [inline linker.ld:1]: Can we please keep the Linker Script and `entry.S` assembly file as simple as possible!
We should get into Rust code as soon as possible.

File: src/asm/entry.S
Comment (Archisman-Mridha) [inline src/asm/entry.S:1]: Please let's have this assembly file as simple as possible and boot into Rust as fast as we can 🙏.

File: src/asm/entry.S [hunk 4]
```diff
@@ -1,20 +1,61 @@
...
+.option    arch, -c       // disable c-extension
+
+.equ STACK_SIZE, 4096
+.equ MPP_M, 3 << 11
+.equ MPP_S, 1 << 11
+
+.section .init
+.global  _start
+
+_start:
```
Comment (Archisman-Mridha) [inline src/asm/entry.S:11]: And you changed the entry symbol name from `_entry` to `start`. And then in the `main.rs` file, you changed the Rust entrypoint function name from `start` to `main`.

Can you revert back please 🙏.

I planned to have a Rust `main` function later, which'll execute CPU specific code - based on whether it's the BSP or some other CPU core.

And the Rust `start` function will do common stuff - like store `SuperVisor` mode in the `mstatus` register, setup trap handler etc...

**Review by Archisman-Mridha — CHANGES_REQUESTED** [review]

File: .cargo/config.toml [hunk 5]
```diff
@@ -2,12 +2,9 @@
 target = "riscv64gc-unknown-none-elf"
 
 [target.riscv64gc-unknown-none-elf]
-rustflags = ["-C", "link-arg=-Tlinker.ld"]
+rustflags = [ "-C", "code-model=medium"]
```
Comment (Archisman-Mridha) [inline .cargo/config.toml:5]: Please explain what's the importance of this.

File: src/asm/entry.S
Comment (bluebrown) [reply src/asm/entry.S:1]: I want to write assembly and learn riscv instructions. Let me do machine in asm, you can do supervisor stuff in rust.

File: build.rs (hunk 3 above)
Comment (bluebrown) [reply build.rs:9]: I know we do, but i want to have a single build file.  Not configs spread around 10 places. The proper place to link an assembly would be build.rs, imo. Inline including a string is just dirty.

File: linker.ld
Comment (bluebrown) [reply linker.ld:1]: I keep removing my own comments too. I do understand we place them as we learn, but it become unwieldy. If you think about it, most of our comments describe very basic things. Something you could look up within 1 second, via web search. Lets try to use comments for more meaningful things. For other stuff, we can always write extra docs or blog posts.

File: .cargo/config.toml (hunk 5 above)
Comment (bluebrown) [reply .cargo/config.toml:5]: This ensures compatible when cross building and linking this. For pure rust, it hasnt been strictly, required. But I have seen errors regarding this a few times in the past weeks.  It says something like "symbols dont fit".

This setting, It aligns the code in s specific way when producing the elf. If you link many things together, you need to make sure they are aligned properly.

Comment (bluebrown) [reply .cargo/config.toml:5]: Here they link to sysv abi https://doc.rust-lang.org/rustc/codegen-options/index.html#code-model

**Review by Archisman-Mridha — APPROVED** [review]
